### PR TITLE
Minify the generated HTML file, if building with -o out.html

### PR DESCRIPTION
**Pull request by juj**
Minify the generated HTML file, if building with -o out.html. Add option -s MINIFY_HTML=0 to allow disabling the minification, in case it is problematic.

The total impact of adding html-minifier into the repository is 2.3MB across 212 files. html-minifier was obtained via running

```
npm install html-minifier -g
```
from

https://github.com/kangax/html-minifier

### CODE REVIEW

**Comment by kripken** [comment]
In general I've made an effort to reduce the size of such bundled libraries, including only the necessary part of the code, and just the readme/changelog/license files but nothing else. Can we do that here too?

**Comment by juj** [comment]
The set of files that are provided here are
```
320    .     
1071   ./LICENSE
12826  ./README.md
11652  ./cli.js
448    ./node_modules
192    ./node_modules/.bin
289    ./node_modules/.bin/he
160    ./node_modules/.bin/he.cmd
315    ./node_modules/.bin/uglifyjs
186    ./node_modules/.bin/uglifyjs.cmd
192    ./node_modules/camel-case
1103   ./node_modules/camel-case/LICENSE
114    ./node_modules/camel-case/camel-case.d.ts
573    ./node_modules/camel-case/camel-case.js
1986   ./node_modules/camel-case/package.json
256    ./node_modules/clean-css
78657  ./node_modules/clean-css/History.md
1062   ./node_modules/clean-css/LICENSE
34845  ./node_modules/clean-css/README.md
41     ./node_modules/clean-css/index.js
288    ./node_modules/clean-css/lib
5351   ./node_modules/clean-css/lib/clean.js
320    ./node_modules/clean-css/lib/optimizer
133    ./node_modules/clean-css/lib/optimizer/hack.js
96     ./node_modules/clean-css/lib/optimizer/level-0
132    ./node_modules/clean-css/lib/optimizer/level-0/optimize.js
320    ./node_modules/clean-css/lib/optimizer/level-1
20543  ./node_modules/clean-css/lib/optimizer/level-1/optimize.js
4491   ./node_modules/clean-css/lib/optimizer/level-1/shorten-hex.js
1408   ./node_modules/clean-css/lib/optimizer/level-1/shorten-hsl.js
450    ./node_modules/clean-css/lib/optimizer/level-1/shorten-rgb.js
529    ./node_modules/clean-css/lib/optimizer/level-1/sort-selectors.js
177    ./node_modules/clean-css/lib/optimizer/level-1/tidy-at-rule.js
691    ./node_modules/clean-css/lib/optimizer/level-1/tidy-block.js
6972   ./node_modules/clean-css/lib/optimizer/level-1/tidy-rules.js
896    ./node_modules/clean-css/lib/optimizer/level-2
21656  ./node_modules/clean-css/lib/optimizer/level-2/break-up.js
10471  ./node_modules/clean-css/lib/optimizer/level-2/can-override.js
772    ./node_modules/clean-css/lib/optimizer/level-2/clone.js
26111  ./node_modules/clean-css/lib/optimizer/level-2/compactable.js
1944   ./node_modules/clean-css/lib/optimizer/level-2/extract-properties.js
316    ./node_modules/clean-css/lib/optimizer/level-2/invalid-property-error.js
6938   ./node_modules/clean-css/lib/optimizer/level-2/is-mergeable.js
2193   ./node_modules/clean-css/lib/optimizer/level-2/merge-adjacent.js
3142   ./node_modules/clean-css/lib/optimizer/level-2/merge-media-queries.js
2924   ./node_modules/clean-css/lib/optimizer/level-2/merge-non-adjacent-by-body.js
2565   ./node_modules/clean-css/lib/optimizer/level-2/merge-non-adjacent-by-selector.js
3848   ./node_modules/clean-css/lib/optimizer/level-2/optimize.js
448    ./node_modules/clean-css/lib/optimizer/level-2/properties
780    ./node_modules/clean-css/lib/optimizer/level-2/properties/every-values-pair.js
953    ./node_modules/clean-css/lib/optimizer/level-2/properties/find-component-in.js
204    ./node_modules/clean-css/lib/optimizer/level-2/properties/has-inherit.js
681    ./node_modules/clean-css/lib/optimizer/level-2/properties/is-component-of.js
256    ./node_modules/clean-css/lib/optimizer/level-2/properties/is-mergeable-shorthand.js
12499  ./node_modules/clean-css/lib/optimizer/level-2/properties/merge-into-shorthands.js
1487   ./node_modules/clean-css/lib/optimizer/level-2/properties/optimize.js
14593  ./node_modules/clean-css/lib/optimizer/level-2/properties/override-properties.js
346    ./node_modules/clean-css/lib/optimizer/level-2/properties/overrides-non-component-shorthand.js
1279   ./node_modules/clean-css/lib/optimizer/level-2/properties/populate-components.js
360    ./node_modules/clean-css/lib/optimizer/level-2/properties/understandable.js
449    ./node_modules/clean-css/lib/optimizer/level-2/properties/vendor-prefixes.js
5368   ./node_modules/clean-css/lib/optimizer/level-2/reduce-non-adjacent.js
619    ./node_modules/clean-css/lib/optimizer/level-2/remove-duplicate-font-at-rules.js
661    ./node_modules/clean-css/lib/optimizer/level-2/remove-duplicate-media-queries.js
978    ./node_modules/clean-css/lib/optimizer/level-2/remove-duplicates.js
6783   ./node_modules/clean-css/lib/optimizer/level-2/remove-unused-at-rules.js
3738   ./node_modules/clean-css/lib/optimizer/level-2/reorderable.js
319    ./node_modules/clean-css/lib/optimizer/level-2/restore-with-components.js
9829   ./node_modules/clean-css/lib/optimizer/level-2/restore.js
12589  ./node_modules/clean-css/lib/optimizer/level-2/restructure.js
596    ./node_modules/clean-css/lib/optimizer/level-2/rules-overlap.js
806    ./node_modules/clean-css/lib/optimizer/level-2/specificities-overlap.js
2482   ./node_modules/clean-css/lib/optimizer/level-2/specificity.js
397    ./node_modules/clean-css/lib/optimizer/level-2/tidy-rule-duplicates.js
244    ./node_modules/clean-css/lib/optimizer/remove-unused.js
1721   ./node_modules/clean-css/lib/optimizer/restore-from-optimizing.js
11111  ./node_modules/clean-css/lib/optimizer/validator.js
4990   ./node_modules/clean-css/lib/optimizer/wrap-for-optimizing.js
384    ./node_modules/clean-css/lib/options
4483   ./node_modules/clean-css/lib/options/compatibility.js
172    ./node_modules/clean-css/lib/options/fetch.js
5135   ./node_modules/clean-css/lib/options/format.js
476    ./node_modules/clean-css/lib/options/inline-request.js
141    ./node_modules/clean-css/lib/options/inline-timeout.js
248    ./node_modules/clean-css/lib/options/inline.js
5563   ./node_modules/clean-css/lib/options/optimization-level.js
151    ./node_modules/clean-css/lib/options/rebase-to.js
129    ./node_modules/clean-css/lib/options/rebase.js
1965   ./node_modules/clean-css/lib/options/rounding-precision.js
512    ./node_modules/clean-css/lib/reader
7930   ./node_modules/clean-css/lib/reader/apply-source-maps.js
854    ./node_modules/clean-css/lib/reader/extract-import-url-and-media.js
1383   ./node_modules/clean-css/lib/reader/input-source-map-tracker.js
2067   ./node_modules/clean-css/lib/reader/is-allowed-resource.js
3830   ./node_modules/clean-css/lib/reader/load-original-sources.js
1867   ./node_modules/clean-css/lib/reader/load-remote-resource.js
173    ./node_modules/clean-css/lib/reader/match-data-uri.js
198    ./node_modules/clean-css/lib/reader/normalize-path.js
13280  ./node_modules/clean-css/lib/reader/read-sources.js
446    ./node_modules/clean-css/lib/reader/rebase-local-map.js
330    ./node_modules/clean-css/lib/reader/rebase-remote-map.js
2588   ./node_modules/clean-css/lib/reader/rebase.js
132    ./node_modules/clean-css/lib/reader/restore-import.js
2701   ./node_modules/clean-css/lib/reader/rewrite-url.js
160    ./node_modules/clean-css/lib/tokenizer
509    ./node_modules/clean-css/lib/tokenizer/marker.js
831    ./node_modules/clean-css/lib/tokenizer/token.js
25119  ./node_modules/clean-css/lib/tokenizer/tokenize.js
416    ./node_modules/clean-css/lib/utils
241    ./node_modules/clean-css/lib/utils/clone-array.js
239    ./node_modules/clean-css/lib/utils/format-position.js
156    ./node_modules/clean-css/lib/utils/has-protocol.js
183    ./node_modules/clean-css/lib/utils/is-data-uri-resource.js
152    ./node_modules/clean-css/lib/utils/is-http-resource.js
157    ./node_modules/clean-css/lib/utils/is-https-resource.js
143    ./node_modules/clean-css/lib/utils/is-import.js
166    ./node_modules/clean-css/lib/utils/is-remote-resource.js
772    ./node_modules/clean-css/lib/utils/natural-compare.js
729    ./node_modules/clean-css/lib/utils/override.js
1113   ./node_modules/clean-css/lib/utils/split.js
192    ./node_modules/clean-css/lib/writer
7344   ./node_modules/clean-css/lib/writer/helpers.js
1111   ./node_modules/clean-css/lib/writer/one-time.js
1303   ./node_modules/clean-css/lib/writer/simple.js
2937   ./node_modules/clean-css/lib/writer/source-maps.js
2030   ./node_modules/clean-css/package.json
256    ./node_modules/commander
10317  ./node_modules/commander/CHANGELOG.md
1098   ./node_modules/commander/LICENSE
12381  ./node_modules/commander/Readme.md
28123  ./node_modules/commander/index.js
1880   ./node_modules/commander/package.json
96     ./node_modules/commander/typings
8401   ./node_modules/commander/typings/index.d.ts
256    ./node_modules/he
1077   ./node_modules/he/LICENSE-MIT.txt
14210  ./node_modules/he/README.md
96     ./node_modules/he/bin
3626   ./node_modules/he/bin/he
100899 ./node_modules/he/he.js
96     ./node_modules/he/man
3102   ./node_modules/he/man/he.1
2153   ./node_modules/he/package.json
224    ./node_modules/lower-case
1103   ./node_modules/lower-case/LICENSE
1359   ./node_modules/lower-case/README.md
90     ./node_modules/lower-case/lower-case.d.ts
1065   ./node_modules/lower-case/lower-case.js
1978   ./node_modules/lower-case/package.json
256    ./node_modules/no-case
1103   ./node_modules/no-case/LICENSE
1629   ./node_modules/no-case/README.md
106    ./node_modules/no-case/no-case.d.ts
1105   ./node_modules/no-case/no-case.js
2060   ./node_modules/no-case/package.json
160    ./node_modules/no-case/vendor
8779   ./node_modules/no-case/vendor/camel-case-regexp.js
11876  ./node_modules/no-case/vendor/camel-case-upper-regexp.js
5073   ./node_modules/no-case/vendor/non-word-regexp.js
224    ./node_modules/param-case
1103   ./node_modules/param-case/LICENSE
1366   ./node_modules/param-case/README.md
1821   ./node_modules/param-case/package.json
90     ./node_modules/param-case/param-case.d.ts
225    ./node_modules/param-case/param-case.js
192    ./node_modules/relateurl
6505   ./node_modules/relateurl/README.md
288    ./node_modules/relateurl/lib
167    ./node_modules/relateurl/lib/constants.js
3433   ./node_modules/relateurl/lib/format.js
1798   ./node_modules/relateurl/lib/index.js
845    ./node_modules/relateurl/lib/options.js
288    ./node_modules/relateurl/lib/parse
366    ./node_modules/relateurl/lib/parse/host.js
759    ./node_modules/relateurl/lib/parse/hrefInfo.js
1100   ./node_modules/relateurl/lib/parse/index.js
1634   ./node_modules/relateurl/lib/parse/path.js
524    ./node_modules/relateurl/lib/parse/port.js
878    ./node_modules/relateurl/lib/parse/query.js
2364   ./node_modules/relateurl/lib/parse/urlstring.js
192    ./node_modules/relateurl/lib/relate
2294   ./node_modules/relateurl/lib/relate/absolutize.js
3549   ./node_modules/relateurl/lib/relate/findRelation.js
287    ./node_modules/relateurl/lib/relate/index.js
1050   ./node_modules/relateurl/lib/relate/relativize.js
160    ./node_modules/relateurl/lib/util
297    ./node_modules/relateurl/lib/util/devlog.js
826    ./node_modules/relateurl/lib/util/object.js
576    ./node_modules/relateurl/lib/util/path.js
1111   ./node_modules/relateurl/license
1821   ./node_modules/relateurl/package.json
320    ./node_modules/source-map
7884   ./node_modules/source-map/CHANGELOG.md
1526   ./node_modules/source-map/LICENSE
24072  ./node_modules/source-map/README.md
192    ./node_modules/source-map/dist
272874 ./node_modules/source-map/dist/source-map.debug.js
106973 ./node_modules/source-map/dist/source-map.js
27111  ./node_modules/source-map/dist/source-map.min.js
257409 ./node_modules/source-map/dist/source-map.min.js.map
384    ./node_modules/source-map/lib
3197   ./node_modules/source-map/lib/array-set.js
4714   ./node_modules/source-map/lib/base64-vlq.js
1540   ./node_modules/source-map/lib/base64.js
4249   ./node_modules/source-map/lib/binary-search.js
2339   ./node_modules/source-map/lib/mapping-list.js
3616   ./node_modules/source-map/lib/quick-sort.js
40562  ./node_modules/source-map/lib/source-map-consumer.js
14356  ./node_modules/source-map/lib/source-map-generator.js
13808  ./node_modules/source-map/lib/source-node.js
12950  ./node_modules/source-map/lib/util.js
4977   ./node_modules/source-map/package.json
3060   ./node_modules/source-map/source-map.d.ts
405    ./node_modules/source-map/source-map.js
256    ./node_modules/uglify-js
1348   ./node_modules/uglify-js/LICENSE
45577  ./node_modules/uglify-js/README.md
96     ./node_modules/uglify-js/bin
14914  ./node_modules/uglify-js/bin/uglifyjs
416    ./node_modules/uglify-js/lib
31555  ./node_modules/uglify-js/lib/ast.js
277008 ./node_modules/uglify-js/lib/compress.js
9519   ./node_modules/uglify-js/lib/minify.js
23055  ./node_modules/uglify-js/lib/mozilla-ast.js
50996  ./node_modules/uglify-js/lib/output.js
59503  ./node_modules/uglify-js/lib/parse.js
8037   ./node_modules/uglify-js/lib/propmangle.js
20024  ./node_modules/uglify-js/lib/scope.js
3919   ./node_modules/uglify-js/lib/sourcemap.js
6888   ./node_modules/uglify-js/lib/transform.js
8482   ./node_modules/uglify-js/lib/utils.js
2292   ./node_modules/uglify-js/package.json
224    ./node_modules/uglify-js/tools
120421 ./node_modules/uglify-js/tools/domprops.json
469    ./node_modules/uglify-js/tools/exit.js
206    ./node_modules/uglify-js/tools/exports.js
2155   ./node_modules/uglify-js/tools/node.js
1640   ./node_modules/uglify-js/tools/props.html
224    ./node_modules/upper-case
1103   ./node_modules/upper-case/LICENSE
1359   ./node_modules/upper-case/README.md
1963   ./node_modules/upper-case/package.json
90     ./node_modules/upper-case/upper-case.d.ts
953    ./node_modules/upper-case/upper-case.js
2904   ./package.json
1033   ./sample-cli-config-file.conf
192    ./src 
45313  ./src/htmlminifier.js
19850  ./src/htmlparser.js
1813   ./src/tokenchain.js
423    ./src/utils.js
```
I don't know of a way to safely shrink this by any meaningful amount.

Tried using rollup to bundle the html-minifier project, but it looks like html-minifier is not rollup compatible, so that was a bit of a dead end.

File: emcc.py [hunk 1]
```diff
@@ -2975,6 +2970,67 @@ def generate_html(target, options, js_target, target_basename,
...
+  # '--remove-empty-elements': removes all elements with empty contents.
+  #                            (Breaks at least browser.test_asm_swapping)
+
+  if options.debug_level >= 2:
+    return
+
+  logger.debug('minifying HTML file ' + filename)
+  size_before = os.path.getsize(filename)
+  cmd = shared.NODE_JS + [shared.path_from_root('third_party', 'html-minifier', 'cli.js'), filename, '-o', filename] + opts
+  logger.debug(str(cmd))
```
Comment (kripken) [inline emcc.py]: this is not needed, `run_process` will log if debugging.

File: src/shell_minimal_runtime.html [hunk 2]
```diff
@@ -4,14 +4,14 @@
 <script>
   var Module = {};
 
-  function script(url, cb) {
+  function s(url, cb) { // Downloads a script file and adds it to DOM
```
Comment (kripken) [inline src/shell_minimal_runtime.html]: did you minify these names manually because the minifier can't?

File: emcc.py (hunk 1 above)
Comment (juj) [reply emcc.py]: check

File: src/shell_minimal_runtime.html (hunk 2 above)
Comment (juj) [reply src/shell_minimal_runtime.html]: Yeah, the minifier does not do minification on top level symbols, only inside functions when it knows there are no external references, so minified that manually.

**Pull request merged**